File: app/src/main/java/com/example/opengl/devourer/Devourer.java
package com.example.opengl.devourer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.library.baseAdapters.BR;

import com.example.opengl.Game;
import com.example.opengl.Position;
import com.example.opengl.PositionInd;
import com.example.opengl.Resource;
import com.example.opengl.ResourcesStorage;
import com.example.opengl.tiles.MineralType;
import com.example.opengl.tiles.TileMap;

import java.nio.FloatBuffer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class Devourer {
    private final TileMap tileMap;
    private final DevourerStructure devourerStructure;
    private final List<Resource> movingResources = Collections.synchronizedList(new ArrayList<>());
    private final ResourcesStorage resourcesStorage = new ResourcesStorage();
    private final Position mainDevourerPos;
    private final float betweenTileCentersX;
    private final float betweenTileCentersY;

    public Devourer(TileMap tileMap, float betweenTileCentersX, float betweenTileCentersY) {
        this.tileMap = tileMap;
        this.betweenTileCentersX = betweenTileCentersX;
        this.betweenTileCentersY = betweenTileCentersY;
        this.devourerStructure = new DevourerStructure();
        devourerStructure.createStructure(tileMap);
        PositionInd mainDevourerPosInd = new PositionInd(tileMap.getMainDevourerTileIndX(), tileMap.getMainDevourerTileIndY());
        mainDevourerPos = calculateDevourerNodePosition(mainDevourerPosInd);
    }

    public DevourerNode getDevourerNode(int x, int y) {
        return devourerStructure.getDevourerNode(x, y);
    }

    public boolean moveResources() {
        if (movingResources.size() > 0) {
            synchronized (movingResources) {
                // Must be in synchronized block
                Iterator<Resource> iterator = movingResources.iterator();
                while (iterator.hasNext()) {
                    Resource resource = iterator.next();
                    if (resource.isMoving) {
                        resource.move(0.1f);
                    }
                    if (resource.isDelivered) {
                        resourcesStorage.addMineral(resource.mineralType, 1);
                        Game.getInstance().showMineralsCounts();
                        iterator.remove();
                    }
                }
            }
            return true;
        } else {
            return false;
        }
    }

    public void placeDevourer(int indX, int indY) {
        if (tileMap.getTile(indX, indY).getMineral() != null) { // is mineral
            startEatingMineral(indX, indY);
        } else {
            tileMap.placeDevourerTile(indX, indY);
            devourerStructure.addDevourerNode(indX, indY, tileMap);
            Game.getInstance().setMessage1("indX:" + indX + " indY:" + indY + " tile:" + tileMap.getTile(indX, indY).getEntity());
        }
    }

    private void startEatingMineral(int indX, int indY) {
        tileMap.placeDevourerTile(indX, indY);
        devourerStructure.addDevourerNode(indX, indY, tileMap);
        DevourerNode devourerNode = devourerStructure.getDevourerNode(indX, indY);
        if (devourerNode != null) {
            tileMap.getTile(indX, indY).getMineral().resourceDepository.startResourceMining(indX, indY, devourerStructure, calculateDevourerNodePosition(new PositionInd(indX, indY)), movingResources);
        }
    }

    public void deleteDevourer(int indX, int indY) {
        tileMap.deleteDevourerTile(indX, indY);
        devourerStructure.removeDevourerNode(indX, indY, tileMap);
        Game.getInstance().setMessage1("indX:" + indX + " indY:" + indY + " tile:" + tileMap.getTile(indX, indY).getBasis());
    }

    public FloatBuffer createRenderDataMoving() {
        synchronized (movingResources) {
            List<Float> resourcesData = new ArrayList<>();
            for (Resource resource : movingResources) {
                Position position = calculateResourcePosition(resource);
                if (position != null) {
                    resourcesData.add(position.x);
                    resourcesData.add(position.y);
                    resourcesData.add(0f); // tile number
                    resourcesData.add(0f); // texture number
                }
            }

            resourcesData.add(mainDevourerPos.x);
            resourcesData.add(mainDevourerPos.y);
            resourcesData.add(0f); // tile number
            resourcesData.add(1f); // texture number main devourer

            //if (resourcesData.size() == 0) return null;
            FloatBuffer result = FloatBuffer.allocate(resourcesData.size());
            for (float fl : resourcesData) {
                result.put(fl);
            }
            return result;
        }
    }

    @NonNull
    private Position calculateDevourerNodePosition(@NonNull PositionInd positionInd) {
        Position result = new Position();
        result.x = positionInd.x * betweenTileCentersX;
        if (positionInd.x % 2 == 0) {
            result.y = positionInd.y * betweenTileCentersY;
        } else {
            result.y = (positionInd.y + 0.5f) * betweenTileCentersY;
        }
        return result;
    }

    @Nullable
    private Position calculateResourcePosition(@NonNull Resource resource) { // interpolation
        Position result = new Position();
        if (resource.nextPosition == null) {
            if (resource.nextPositionInd != null) {
                resource.nextPosition = calculateDevourerNodePosition(resource.nextPositionInd);
            } else {
                return null;
            }
        }
        result.x = resource.startPosition.x + (resource.nextPosition.x - resource.startPosition.x) * resource.distance;
        result.y = resource.startPosition.y + (resource.nextPosition.y - resource.startPosition.y) * resource.distance;
        return result;
    }

    public int getMineralCount(MineralType mineralType) {
        return resourcesStorage.getMineralCount(mineralType);
    }
}

File: app/src/main/java/com/example/opengl/devourer/Devourer.kt
package com.example.opengl.devourer

import com.example.opengl.Game
import com.example.opengl.Position
import com.example.opengl.PositionInd
import com.example.opengl.Resource
import com.example.opengl.ResourcesStorage
import com.example.opengl.tiles.MineralType
import com.example.opengl.tiles.TileMap
import java.nio.FloatBuffer
import java.util.Collections
import java.util.Timer
import java.util.TimerTask

class Devourer(
    private val tileMap: TileMap,
    private val betweenTileCentersX: Float,
    private val betweenTileCentersY: Float,
) {
    enum class DevourerState {
        NORMAL, EATING
    }

    private val devourerStructure: DevourerStructure = DevourerStructure()
    private val movingResources = Collections.synchronizedList(ArrayList<Resource>())
    private val resourcesStorage = ResourcesStorage()
    private val mainDevourerPos: Position
    private val animationTimer: Timer
    private var animationPhase = 0
    private var devourerState = DevourerState.NORMAL

    init {
        devourerStructure.createStructure(tileMap)
        val mainDevourerPosInd =
            PositionInd(tileMap.mainDevourerTileIndX, tileMap.mainDevourerTileIndY)
        mainDevourerPos = calculateDevourerNodePosition(mainDevourerPosInd)
        animationTimer = Timer()
        animationTimer.schedule(object : TimerTask() {
            override fun run() {
                calculateDevourerAnimationPhase()
            }
        }, 0, 33)
    }

    private fun calculateDevourerAnimationPhase() {
        when (devourerState) {
            DevourerState.NORMAL -> if (animationPhase == NORMAL_PHASES.size - 1) {
                animationPhase = 0
            } else {
                animationPhase++
            }

            DevourerState.EATING -> if (animationPhase == EATING_PHASES.size - 1) {
                animationPhase = 0
                setDevourerState(DevourerState.NORMAL)
            } else {
                animationPhase++
            }
        }
    }

    private val currentAnimationSpriteNumber: Float
        get() {
            var result = 0
            result = when (devourerState) {
                DevourerState.NORMAL -> NORMAL_PHASES[animationPhase]
                DevourerState.EATING -> EATING_PHASES[animationPhase]
            }
            return result.toFloat()
        }

    private fun setDevourerState(newDevourerState: DevourerState) {
        if (devourerState != newDevourerState) {
            devourerState = newDevourerState
            animationPhase = 0
        }
    }

    fun getDevourerNode(x: Int, y: Int): DevourerNode? {
        return devourerStructure.getDevourerNode(x, y)
    }

    fun moveResources(): Boolean {
        return if (movingResources.size > 0) {
            synchronized(movingResources) {
                // Must be in synchronized block
                val iterator = movingResources.iterator()
                while (iterator.hasNext()) {
                    val resource = iterator.next()
                    if (resource.isMoving) {
                        resource.move(MINERAL_SPEED)
                    } else {
                        resource.tryMove(MINERAL_SPEED)                    }
                    if (resource.isDelivered) {
                        resourcesStorage.addMineral(resource.mineralType, 1)
                        setDevourerState(DevourerState.EATING)
                        Game.instance.showMineralsCounts()
                        iterator.remove()
                    }
                }
            }
            true
        } else {
            false
        }
    }

    fun placeDevourer(indX: Int, indY: Int) {
        if (tileMap.getTile(indX, indY)!!.mineral != null) { // is mineral
            startEatingMineral(indX, indY)
        } else {
            tileMap.placeDevourerTile(indX, indY)
            devourerStructure.addDevourerNode(indX, indY, tileMap)
            Game.instance.setMessage1(
                "indX:" + indX + " indY:" + indY + " tile:" + tileMap.getTile(
                    indX,
                    indY
                )!!.entity
            )
        }
    }

    private fun startEatingMineral(indX: Int, indY: Int) {
        tileMap.placeDevourerTile(indX, indY)
        devourerStructure.addDevourerNode(indX, indY, tileMap)
        val devourerNode = devourerStructure.getDevourerNode(indX, indY)
        if (devourerNode != null) {
            tileMap.getTile(indX, indY)!!.mineral?.resourceDepository?.startResourceMining(
                indX,
                indY,
                devourerStructure,
                calculateDevourerNodePosition(PositionInd(indX, indY)),
                movingResources
            )
        }
    }

    fun deleteDevourer(indX: Int, indY: Int) {
        tileMap.deleteDevourerTile(indX, indY)
        devourerStructure.removeDevourerNode(indX, indY, tileMap)
        Game.instance.setMessage1(
            "indX:" + indX + " indY:" + indY + " tile:" + tileMap.getTile(
                indX,
                indY
            )!!.basis
        )
    }

    fun createRenderDataMoving(): FloatBuffer {
        synchronized(movingResources) {
            val resourcesData: MutableList<Float> = ArrayList()
            for (resource in movingResources) {
                val position = calculateResourcePosition(resource)
                if (position != null) {
                    resourcesData.add(position.x)
                    resourcesData.add(position.y)
                    resourcesData.add(0f) // tile number
                    resourcesData.add(0f) // texture number
                }
            }
            resourcesData.add(mainDevourerPos.x)
            resourcesData.add(mainDevourerPos.y)
            resourcesData.add(currentAnimationSpriteNumber) // sprite number
            resourcesData.add(1f) // texture number main devourer

            //if (resourcesData.size() == 0) return null;
            val result = FloatBuffer.allocate(resourcesData.size)
            for (fl in resourcesData) {
                result.put(fl)
            }
            return result
        }
    }

    private fun calculateDevourerNodePosition(positionInd: PositionInd): Position {
        val result = Position()
        result.x = positionInd.x * betweenTileCentersX
        if (positionInd.x % 2 == 0) {
            result.y = positionInd.y * betweenTileCentersY
        } else {
            result.y = (positionInd.y + 0.5f) * betweenTileCentersY
        }
        return result
    }

    private fun calculateResourcePosition(resource: Resource): Position? { // interpolation
        val result = Position()
        if (resource.nextPosition == null) {
            if (resource.nextPositionInd != null) {
                resource.nextPosition = calculateDevourerNodePosition(resource.nextPositionInd!!)
            } else {
                return null
            }
        }
        result.x =
            resource.startPosition!!.x + (resource.nextPosition!!.x - resource.startPosition!!.x) * resource.distance
        result.y =
            resource.startPosition!!.y + (resource.nextPosition!!.y - resource.startPosition!!.y) * resource.distance
        return result
    }

    fun getMineralCount(mineralType: MineralType?): Int {
        return resourcesStorage.getMineralCount(mineralType)
    }

    companion object {
        private val EATING_PHASES = intArrayOf(
            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22
        )
        private val NORMAL_PHASES = intArrayOf(
            0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 22, 22, 22, 22, 22, 2, 2, 2, 1, 1, 1
        )
        private val MINERAL_SPEED : Float = 0.1f
    }
}